File: backend-legacy/app/utils/api_helpers.py
"""
API Helper Utilities
Common helpers for API endpoints
"""
import json
import logging
from typing import Dict, Any
from flask import jsonify
from flask_jwt_extended import get_jwt_identity

logger = logging.getLogger(__name__)


def get_current_user_id() -> int:
    """
    Helper to get and convert current user ID from JWT.
    
    Returns:
        User ID as integer
        
    Raises:
        ValueError: If token format is invalid
    """
    current_user_id = get_jwt_identity()
    try:
        return int(current_user_id) if isinstance(current_user_id, str) else current_user_id
    except (ValueError, TypeError):
        raise ValueError('Invalid token format')


def format_sse_message(event: str, data: Dict[str, Any]) -> str:
    """
    Format a Server-Sent Events message.
    
    Args:
        event: The event type
        data: The data payload
        
    Returns:
        Formatted SSE message string
    """
    return f"event: {event}\ndata: {json.dumps(data)}\n\n"


def create_error_response(message: str, status_code: int = 400, details: Any = None) -> tuple:
    """
    Create a standardized error response.
    
    Args:
        message: Error message
        status_code: HTTP status code
        details: Optional additional details
        
    Returns:
        Tuple of (response, status_code)
    """
    error_obj = {'error': {'message': message}}
    if details:
        error_obj['error']['details'] = details
    return jsonify(error_obj), status_code


def cleanup_presentation_on_error(presentation_id: int) -> None:
    """
    Delete a presentation when an error occurs during generation.
    
    Args:
        presentation_id: ID of the presentation to delete
    """
    try:
        from app.models import Presentation, db
        pres = db.session.get(Presentation, presentation_id)
        if pres:
            db.session.delete(pres)
            db.session.commit()
            logger.info(f"Cleaned up failed presentation {presentation_id}")
    except Exception as e:
        logger.error(f"Error cleaning up presentation {presentation_id}: {e}")


def map_error_to_status_code(error_message: str) -> int:
    """
    Map error messages to appropriate HTTP status codes.
    
    Args:
        error_message: The error message
        
    Returns:
        Appropriate HTTP status code
    """
    error_lower = error_message.lower()
    
    if 'not found' in error_lower:
        return 404
    elif 'unauthorized' in error_lower or 'forbidden' in error_lower:
        return 403
    elif 'insufficient tokens' in error_lower or 'payment' in error_lower:
        return 402  # Payment Required
    elif 'conflict' in error_lower or 'already exists' in error_lower:
        return 409
    else:
        return 400

File: apps/backend/src/utils/api-helpers.ts
/**
 * API Helper Utilities
 * Common helpers for API endpoints
 */

/**
 * Format a Server-Sent Events message.
 */
export function formatSSEMessage(event: string, data: Record<string, any>): string {
  return `event: ${event}\ndata: ${JSON.stringify(data)}\n\n`;
}

/**
 * Create a standardized error response.
 */
export function createErrorResponse(
  message: string,
  statusCode = 400,
  details?: any
): { error: { message: string; details?: any } } {
  const errorObj: { error: { message: string; details?: any } } = {
    error: { message },
  };

  if (details) {
    errorObj.error.details = details;
  }

  return errorObj;
}

/**
 * Map error messages to appropriate HTTP status codes.
 */
export function mapErrorToStatusCode(errorMessage: string): number {
  const errorLower = errorMessage.toLowerCase();

  if (errorLower.includes('not found')) {
    return 404;
  } else if (errorLower.includes('unauthorized') || errorLower.includes('forbidden')) {
    return 403;
  } else if (errorLower.includes('insufficient tokens') || errorLower.includes('payment')) {
    return 402; // Payment Required
  } else if (errorLower.includes('conflict') || errorLower.includes('already exists')) {
    return 409;
  } else {
    return 400;
  }
}

/**
 * Validate request parameters and throw descriptive errors
 */
export function validateRequiredFields(data: Record<string, any>, requiredFields: string[]): void {
  const missingFields = requiredFields.filter(
    (field) => data[field] === undefined || data[field] === null || data[field] === ''
  );

  if (missingFields.length > 0) {
    throw new Error(`Missing required fields: ${missingFields.join(', ')}`);
  }
}

/**
 * Sanitize and validate slide count
 */
export function validateSlideCount(slideCount: any): number {
  const count = Number.parseInt(slideCount);

  if (isNaN(count) || count < 1 || count > 50) {
    throw new Error('Slide count must be between 1 and 50');
  }

  return count;
}

/**
 * Validate detail level parameter
 */
export function validateDetailLevel(detailLevel: string): string {
  const validLevels = ['brief', 'concise', 'balanced', 'detailed', 'comprehensive'];

  if (!validLevels.includes(detailLevel)) {
    throw new Error(`Invalid detail level. Must be one of: ${validLevels.join(', ')}`);
  }

  return detailLevel;
}

/**
 * Validate tonality parameter
 */
export function validateTonality(tonality: string): string {
  const validTonalities = ['professional', 'casual', 'enthusiastic', 'persuasive'];

  if (!validTonalities.includes(tonality)) {
    throw new Error(`Invalid tonality. Must be one of: ${validTonalities.join(', ')}`);
  }

  return tonality;
}

/**
 * Clean and validate topic string
 */
export function validateTopic(topic: string): string {
  if (!topic || typeof topic !== 'string') {
    throw new Error('Topic is required and must be a string');
  }

  const cleanTopic = topic.trim();

  if (cleanTopic.length < 1 || cleanTopic.length > 500) {
    throw new Error('Topic must be between 1 and 500 characters');
  }

  return cleanTopic;
}

/**
 * Extract user ID from JWT token (placeholder - implement based on your auth system)
 */
export function getCurrentUserId(authHeader?: string): number {
  // This is a placeholder implementation
  // In a real implementation, you would decode and validate the JWT token
  if (!authHeader || !authHeader.startsWith('Bearer ')) {
    throw new Error('Invalid authorization header');
  }

  // For now, return a mock user ID
  // Replace this with actual JWT decoding logic
  return 1;
}

/**
 * Generate a unique presentation ID
 */
export function generatePresentationId(): string {
  return `pres_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
}

/**
 * Format presentation data for API response
 */
export function formatPresentationResponse(presentation: any): any {
  return {
    id: presentation.id,
    title: presentation.title,
    prompt: presentation.prompt,
    slides_data: presentation.slides_data,
    created_at: presentation.created_at,
    updated_at: presentation.updated_at,
    user_id: presentation.user_id,
  };
}

/**
 * Calculate estimated processing time based on slide count and complexity
 */
export function estimateProcessingTime(slideCount: number, detailLevel: string): number {
  const baseTimePerSlide = 2; // seconds

  const detailMultipliers: Record<string, number> = {
    brief: 0.7,
    concise: 0.9,
    balanced: 1.0,
    detailed: 1.5,
    comprehensive: 2.0,
  };

  const multiplier = detailMultipliers[detailLevel] || 1.0;
  return Math.ceil(slideCount * baseTimePerSlide * multiplier);
}
